Refactor `build-thread` for Typechecking, readability, modernity

lyzadanger opened the pull request:
This is a foundational part of https://github.com/hypothesis/client/milestone/1

This PR refactors the `build-thread` module to:

* Add and correct typechecking and fix issues arising from it
* Clarify the code for readability
* Modernize to ES6 syntax
* Add comments for the next developer

lyzadanger commented on src/sidebar/build-thread.js:
```diff
@@ -2,107 +2,106 @@
+ * @return {string}
  */
-function id(annotation) {
+function annotationId(annotation) {
```
Right now this module has no dependencies, which is nice, but I feel like this metadata utility function would ideally be in a shared place.

lyzadanger commented on src/sidebar/build-thread.js:
```diff
@@ -114,49 +113,44 @@ function setParentID(threads, id, parents) {
+    const id = annotationId(annotation);
+    threads[id] = {
+      ...DEFAULT_THREAD_STATE,
+      ...{ children: [], annotation, id },
```
It's important to note that `children: []` cannot be part of `DEFAULT_THREAD_STATE` as all derived objects would share the same Array reference. Hilarity ensues.

lyzadanger commented on src/sidebar/build-thread.js:
```diff
@@ -305,86 +297,99 @@ const defaultOpts = {
-    // If the thread was explicitly expanded or collapsed, respect that option
-    if (opts.expanded.hasOwnProperty(id)) {
-      return Object.assign({}, thread, { collapsed: !opts.expanded[id] });
+  // Remove top-level threads which contain no visible annotations
```
I've tried to hoist actions that _remove_ annotations/threads higher in the function but obviously this one depends on derived visibility...

lyzadanger commented on src/sidebar/build-thread.js:
```diff
@@ -305,86 +297,99 @@ const defaultOpts = {
+      collapsed: thread.collapsed,
+    };
+
+    if (hasHighlights) {
```
I've been careful to retain the highlight-related logic here, but it is not presently _applied_ in the application.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -114,49 +113,44 @@ function setParentID(threads, id, parents) {
+    const id = annotationId(annotation);
+    threads[id] = {
+      ...DEFAULT_THREAD_STATE,
+      ...{ children: [], annotation, id },
```
You don't need to use object spread here, since you're just adding a static set of properties to an object literal. IOW - `{ ...DEFAULT_THREAD_STATE, children: [], annotation, id }` would be more efficient.

lyzadanger replied:
My brain must have been on some sort of special "definitely spread everything" trip.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -2,107 +2,106 @@
+    return true;
+  }
+  return hasPathToRoot(threads, id, threads[ancestorId].parent);
+}
```
In thinking through this I found a test case which results in an infinite loop on `master` and infinite recursion in `hasPathToRoot` on this branch:

```js
    it('handles annotations that have their own ID in their `references` list', () => {
      const fixture = [
        { id: '1', references: ['1'] },
        { id: '2', references: ['1'] },
      ];
      const thread = createThread(fixture);
      assert.deepEqual(thread, []);
    });
```

robertknight replied:
If the client ever encountered the annotations from this fixture that would indicate that something has gone elsewhere in the Hypothesis system, but I don't know if anything actually prevents such annotations from being created on the server.

lyzadanger replied:
I have a fix for this, but just FYI that I've updated the test assert to:

```
      assert.deepEqual(thread, [
        {
          annotation: {
            id: '1',
            references: ['1'],
          },
          children: [
            {
              annotation: {
                id: '2',
                references: ['1'],
              },
              children: [],
            },
          ],
        },
      ]);
```

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -165,10 +159,11 @@ function threadAnnotations(annotations) {
  *
  * @param {Thread} thread
  * @param {(Thread) => Thread} mapFn
```
```suggestion
 * @param {(t: Thread) => Thread} mapFn
```

This is a mistake in the previous code, but `(Thread) => Thread` is interpreted by TS as a function that takes an argument of _any_ type _called_ `Thread` and returns a `Thread`.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -203,96 +198,93 @@ function sort(threads, compareFn) {
-  return Object.assign({}, thread, {
-    children: sort(children, compareFn),
-  });
+  return { ...thread, ...{ children: sort(children, compareFn) } };
```
You don't need to use object spread here to add the `children` property.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -203,96 +198,93 @@ function sort(threads, compareFn) {
 
 /**
- * Return a copy of @p thread with the `replyCount` and `depth` properties
+ * Return a copy of thread with the `replyCount` and `depth` properties
```
```suggestion
 * Return a copy of `thread` with the `replyCount` and `depth` properties
```

Make it a bit clearer that `thread` refers to a parameter here.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -203,96 +198,93 @@ function sort(threads, compareFn) {
+  return {
+    ...thread,
+    ...{
+      children,
```
Object spread is not needed here for the `children`, `depth` and `replyCount` properties.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -203,96 +198,93 @@ function sort(threads, compareFn) {
+ * @prop {(t: Thread) => boolean} [threadFilterFn] - Predicate function that
+ *       returns `true` if the annotation should be included in the thread tree
+ * @prop {Object.<string, boolean>} [expanded] - Map of thread id => expansion state
+ * @prop {Object} [highlighted] - List of ids of annotations that are highlighted
```
```suggestion
 * @prop {string[]} [highlighted] - List of ids of annotations that are highlighted
```

lyzadanger replied:
This reminds me: we should conventionlize on either `{Array<type>}` or `{type[]}` perhaps?

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -305,86 +297,99 @@ const defaultOpts = {
-        thread.annotation &&
-        shouldShowThread(thread.annotation),
-    });
+    return { ...thread, ...{ visible: threadIsVisible } };
```
Object spread is not needed here to add the `visible` property.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -305,86 +297,99 @@ const defaultOpts = {
+  const opts = { ...defaultOpts, ...options };
+
+  const annotationsFiltered = !!opts.filterFn;
+  const threadsFiltered = !!opts.threadFilterFn;
```
The `threadsFiltered`, `hasSelection` and `hasHighlights` variables are used only once so I think it would be a marginally simpler just to replace the associated `if (...)` conditions with their initializer, minus the `!!` to-boolean coercion.

lyzadanger replied:
I kind of like this the way it is as it is an explicit set of factors that affect state all in one place. It makes the _intent_ of certain logic easier to understand, IMHO. For example, a later expression:

        const hasUnfilteredChildren =
          annotationsFiltered && hasVisibleChildren(thread);

It's (marginally :) ) easier to understand the intent here vs.

        const hasUnfilteredChildren =
          opts.filterFn && hasVisibleChildren(thread);

I know that `annotationsFiltered` wasn't one of the references you called out here, but I like the completeness of being explicit about what it means for threads to be filtered, if we're defining what it means for annotations to be filtered.

The code I started out with also wasn't clear about delineating between annotation filtering and thread filtering, which I think this helps define.

robertknight commented on src/sidebar/build-thread.js:
```diff
@@ -305,86 +297,99 @@ const defaultOpts = {
+ * (an exception is made if that annotation's thead has been forced visible by
+ * a user).
+ *
  * @param {Array<Annotation>} annotations - A list of annotations and replies
```
```suggestion
 * @param {Annotation[]} annotations - A list of annotations and replies
```

While you're improving types here, I suggest for consistency we should use `T[]` rather than `Array<T>` except for the cases where `T` is some expression that can't be used with the shorter syntax.

lyzadanger replied:
Ah! I had just commented about this above! Great minds...

lyzadanger commented:
* Object spread...diarrhea...cleaned up
* Fixed potential infinite loop if immediate parent is self-referential
* Updated TS per conventions

Pinging for re-review!
